{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "*Pipeline Created by Justin Mumma - Pravetoni Lab, UW Medicine*\n",
        "\n",
        "*Under Construction. Last updated: 02-27-2026*\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "UDQLuqGJhsgY"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setup Instructions\n",
        "- Download legacy .pdb file for mAb + ligand complex (such as 7U64 for fentanyl).\n",
        "\n",
        "**Important:**\n",
        "- you must remove all water molecules, ions, and the assymetric unit from the .pdb structure *(using PyMol)*\n",
        "---\n",
        "\n",
        "#### How to clean and Structure the .pdb file:\n",
        "- in PyMol command line, strip away unnecessary artifacts by typing:\n",
        "\n",
        "  - `remove solvent` (deletes all water)\n",
        "  - `remove inorg` (deletes all salts/ions)\n",
        "\n",
        "- Isolate a single Fab-Fentanyl complex.\n",
        "  - 7U64 usually has Chains A & B (one Fab) and Chains C & D (the second Fab), along with two fentanyl ligands.\n",
        "  - Identify which fentanyl is bound to Chains A & B.\n",
        "  - Delete Chains C, D, and the second fentanyl.\n",
        "\n",
        "- Renaming Chains (**important**):\n",
        "  - Colab script's contig (H1-97/.../L1-214) specifically looks for chain letters H and L. If 7U64 uses A and B, the script will crash.\n",
        "  - Determine which chain is Heavy and which is Light (usually Heavy is the longer one, around 220 amino acids).\n",
        "  - In PyMOL, use the alter command to rename them:\n",
        "\n",
        "    `alter chain A, chain=\"H\"`\n",
        "\n",
        "    `alter chain B, chain=\"L\"`\n",
        "  - Rename the fentanyl chain to something unique, like `chain=\"Z\"`\n",
        "- Save this cleaned complex as `fent_mAb_7U64_clean.pdb`\n",
        "\n",
        "  *(File > Export Molecule > Save as PDB)*,\n",
        "  \n",
        "  and upload to Colab\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "TI-SuTKC5wz2"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# RFDiffusion Partial Diffusion / Motif Scaffolding\n",
        "\n",
        "\n",
        "*   Input your existing antibody structure but tell RFDiffusion to \"mask\" or delete specific CDR loops. u]Usually the heavy chain CDR3, which is most responsible for binding.\n",
        "*    Provide the fentanyl molecule coordinates, keep the rest of the antibody rigid, and ask RFDiffusion to generate new backbone shapes only for that missing loop to better cradle the fentanyl.\n",
        "\n",
        "\n",
        "*   **Next Step:** You pass those new backbone structures to LigandMPNN to generate the amino acid sequences for those new loops\n",
        "\n"
      ],
      "metadata": {
        "id": "jEjeFDUQf-ZS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title setup **RFdiffusion** (~3min)\n",
        "%%time\n",
        "import os, time, signal\n",
        "import sys, random, string, re\n",
        "if not os.path.isdir(\"params\"):\n",
        "  os.system(\"apt-get install aria2\")\n",
        "  os.system(\"mkdir params\")\n",
        "  # send param download into background\n",
        "  os.system(\"(\\\n",
        "  aria2c -q -x 16 https://files.ipd.uw.edu/krypton/schedules.zip; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/6f5902ac237024bdd0c176cb93063dc4/Base_ckpt.pt; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/e29311f6f1bf1af907f9ef9f44b8328b/Complex_base_ckpt.pt; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/f572d396fae9206628714fb2ce00f72e/Complex_beta_ckpt.pt; \\\n",
        "  aria2c -q -x 16 https://storage.googleapis.com/alphafold/alphafold_params_2022-12-06.tar; \\\n",
        "  tar -xf alphafold_params_2022-12-06.tar -C params; \\\n",
        "  touch params/done.txt) &\")\n",
        "\n",
        "if not os.path.isdir(\"RFdiffusion\"):\n",
        "  print(\"installing RFdiffusion...\")\n",
        "  os.system(\"git clone https://github.com/sokrypton/RFdiffusion.git\")\n",
        "  os.system(\"pip install jedi omegaconf hydra-core icecream pyrsistent pynvml decorator\")\n",
        "  os.system(\"pip install git+https://github.com/NVIDIA/dllogger#egg=dllogger\")\n",
        "  # 17Mar2024: adding --no-dependencies to avoid installing nvidia-cuda-* dependencies\n",
        "  # 25Aug2025: updating dgi install to work with latest pytorch\n",
        "  os.system(\"pip install --no-dependencies dgl -f https://data.dgl.ai/wheels/torch-2.4/cu124/repo.html\")\n",
        "  os.system(\"pip install --no-dependencies e3nn==0.5.5 opt_einsum_fx\")\n",
        "  os.system(\"cd RFdiffusion/env/SE3Transformer; pip install .\")\n",
        "  os.system(\"wget -qnc https://files.ipd.uw.edu/krypton/ananas\")\n",
        "  os.system(\"chmod +x ananas\")\n",
        "\n",
        "if not os.path.isdir(\"colabdesign\"):\n",
        "  print(\"installing ColabDesign...\")\n",
        "  os.system(\"pip -q install git+https://github.com/sokrypton/ColabDesign.git\")\n",
        "  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/colabdesign colabdesign\")\n",
        "\n",
        "if not os.path.isdir(\"RFdiffusion/models\"):\n",
        "  print(\"downloading RFdiffusion params...\")\n",
        "  os.system(\"mkdir RFdiffusion/models\")\n",
        "  models = [\"Base_ckpt.pt\",\"Complex_base_ckpt.pt\",\"Complex_beta_ckpt.pt\"]\n",
        "  for m in models:\n",
        "    while os.path.isfile(f\"{m}.aria2\"):\n",
        "      time.sleep(5)\n",
        "  os.system(f\"mv {' '.join(models)} RFdiffusion/models\")\n",
        "  os.system(\"unzip schedules.zip; rm schedules.zip\")\n",
        "\n",
        "if 'RFdiffusion' not in sys.path:\n",
        "  os.environ[\"DGLBACKEND\"] = \"pytorch\"\n",
        "  sys.path.append('RFdiffusion')\n",
        "\n",
        "from google.colab import files\n",
        "import json\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "from IPython.display import display, HTML\n",
        "import ipywidgets as widgets\n",
        "import py3Dmol\n",
        "\n",
        "from inference.utils import parse_pdb\n",
        "from colabdesign.rf.utils import get_ca\n",
        "from colabdesign.rf.utils import fix_contigs, fix_partial_contigs, fix_pdb, sym_it\n",
        "from colabdesign.shared.protein import pdb_to_string\n",
        "from colabdesign.shared.plot import plot_pseudo_3D\n",
        "\n",
        "def get_pdb(pdb_code=None):\n",
        "  if pdb_code is None or pdb_code == \"\":\n",
        "    upload_dict = files.upload()\n",
        "    pdb_string = upload_dict[list(upload_dict.keys())[0]]\n",
        "    with open(\"tmp.pdb\",\"wb\") as out: out.write(pdb_string)\n",
        "    return \"tmp.pdb\"\n",
        "  elif os.path.isfile(pdb_code):\n",
        "    return pdb_code\n",
        "  elif len(pdb_code) == 4:\n",
        "    if not os.path.isfile(f\"{pdb_code}.pdb1\"):\n",
        "      os.system(f\"wget -qnc https://files.rcsb.org/download/{pdb_code}.pdb1.gz\")\n",
        "      os.system(f\"gunzip {pdb_code}.pdb1.gz\")\n",
        "    return f\"{pdb_code}.pdb1\"\n",
        "  else:\n",
        "    os.system(f\"wget -qnc https://alphafold.ebi.ac.uk/files/AF-{pdb_code}-F1-model_v3.pdb\")\n",
        "    return f\"AF-{pdb_code}-F1-model_v3.pdb\"\n",
        "\n",
        "def run_ananas(pdb_str, path, sym=None):\n",
        "  pdb_filename = f\"outputs/{path}/ananas_input.pdb\"\n",
        "  out_filename = f\"outputs/{path}/ananas.json\"\n",
        "  with open(pdb_filename,\"w\") as handle:\n",
        "    handle.write(pdb_str)\n",
        "\n",
        "  cmd = f\"./ananas {pdb_filename} -u -j {out_filename}\"\n",
        "  if sym is None: os.system(cmd)\n",
        "  else: os.system(f\"{cmd} {sym}\")\n",
        "\n",
        "  # parse results\n",
        "  try:\n",
        "    out = json.loads(open(out_filename,\"r\").read())\n",
        "    results,AU = out[0], out[-1][\"AU\"]\n",
        "    group = AU[\"group\"]\n",
        "    chains = AU[\"chain names\"]\n",
        "    rmsd = results[\"Average_RMSD\"]\n",
        "    print(f\"AnAnaS detected {group} symmetry at RMSD:{rmsd:.3}\")\n",
        "\n",
        "    C = np.array(results['transforms'][0]['CENTER'])\n",
        "    A = [np.array(t[\"AXIS\"]) for t in results['transforms']]\n",
        "\n",
        "    # apply symmetry and filter to the asymmetric unit\n",
        "    new_lines = []\n",
        "    for line in pdb_str.split(\"\\n\"):\n",
        "      if line.startswith(\"ATOM\"):\n",
        "        chain = line[21:22]\n",
        "        if chain in chains:\n",
        "          x = np.array([float(line[i:(i+8)]) for i in [30,38,46]])\n",
        "          if group[0] == \"c\":\n",
        "            x = sym_it(x,C,A[0])\n",
        "          if group[0] == \"d\":\n",
        "            x = sym_it(x,C,A[1],A[0])\n",
        "          coord_str = \"\".join([\"{:8.3f}\".format(a) for a in x])\n",
        "          new_lines.append(line[:30]+coord_str+line[54:])\n",
        "      else:\n",
        "        new_lines.append(line)\n",
        "    return results, \"\\n\".join(new_lines)\n",
        "\n",
        "  except:\n",
        "    return None, pdb_str\n",
        "\n",
        "def run(command, steps, num_designs=1, visual=\"none\"):\n",
        "\n",
        "  def run_command_and_get_pid(command):\n",
        "    pid_file = '/dev/shm/pid'\n",
        "    os.system(f'nohup {command} & echo $! > {pid_file}')\n",
        "    with open(pid_file, 'r') as f:\n",
        "      pid = int(f.read().strip())\n",
        "    os.remove(pid_file)\n",
        "    return pid\n",
        "  def is_process_running(pid):\n",
        "    try:\n",
        "      os.kill(pid, 0)\n",
        "    except OSError:\n",
        "      return False\n",
        "    else:\n",
        "      return True\n",
        "\n",
        "  run_output = widgets.Output()\n",
        "  progress = widgets.FloatProgress(min=0, max=1, description='running', bar_style='info')\n",
        "  display(widgets.VBox([progress, run_output]))\n",
        "\n",
        "  # clear previous run\n",
        "  for n in range(steps):\n",
        "    if os.path.isfile(f\"/dev/shm/{n}.pdb\"):\n",
        "      os.remove(f\"/dev/shm/{n}.pdb\")\n",
        "\n",
        "  pid = run_command_and_get_pid(command)\n",
        "  try:\n",
        "    fail = False\n",
        "    for _ in range(num_designs):\n",
        "\n",
        "      # for each step check if output generated\n",
        "      for n in range(steps):\n",
        "        wait = True\n",
        "        while wait and not fail:\n",
        "          time.sleep(0.1)\n",
        "          if os.path.isfile(f\"/dev/shm/{n}.pdb\"):\n",
        "            pdb_str = open(f\"/dev/shm/{n}.pdb\").read()\n",
        "            if pdb_str[-3:] == \"TER\":\n",
        "              wait = False\n",
        "            elif not is_process_running(pid):\n",
        "              fail = True\n",
        "          elif not is_process_running(pid):\n",
        "            fail = True\n",
        "\n",
        "        if fail:\n",
        "          progress.bar_style = 'danger'\n",
        "          progress.description = \"failed\"\n",
        "          break\n",
        "\n",
        "        else:\n",
        "          progress.value = (n+1) / steps\n",
        "          if visual != \"none\":\n",
        "            with run_output:\n",
        "              run_output.clear_output(wait=True)\n",
        "              if visual == \"image\":\n",
        "                xyz, bfact = get_ca(f\"/dev/shm/{n}.pdb\", get_bfact=True)\n",
        "                fig = plt.figure()\n",
        "                fig.set_dpi(100);fig.set_figwidth(6);fig.set_figheight(6)\n",
        "                ax1 = fig.add_subplot(111);ax1.set_xticks([]);ax1.set_yticks([])\n",
        "                plot_pseudo_3D(xyz, c=bfact, cmin=0.5, cmax=0.9, ax=ax1)\n",
        "                plt.show()\n",
        "              if visual == \"interactive\":\n",
        "                view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
        "                view.addModel(pdb_str,'pdb')\n",
        "                view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':0.5,'max':0.9}}})\n",
        "                view.zoomTo()\n",
        "                view.show()\n",
        "        if os.path.exists(f\"/dev/shm/{n}.pdb\"):\n",
        "          os.remove(f\"/dev/shm/{n}.pdb\")\n",
        "      if fail:\n",
        "        progress.bar_style = 'danger'\n",
        "        progress.description = \"failed\"\n",
        "        break\n",
        "\n",
        "    while is_process_running(pid):\n",
        "      time.sleep(0.1)\n",
        "\n",
        "  except KeyboardInterrupt:\n",
        "    os.kill(pid, signal.SIGTERM)\n",
        "    progress.bar_style = 'danger'\n",
        "    progress.description = \"stopped\"\n",
        "\n",
        "def run_diffusion(contigs, path, pdb=None, iterations=50,\n",
        "                  symmetry=\"none\", order=1, hotspot=None,\n",
        "                  chains=None, add_potential=False, partial_T=\"auto\",\n",
        "                  num_designs=1, use_beta_model=False, visual=\"none\"):\n",
        "\n",
        "  full_path = f\"outputs/{path}\"\n",
        "  os.makedirs(full_path, exist_ok=True)\n",
        "  opts = [f\"inference.output_prefix={full_path}\",\n",
        "          f\"inference.num_designs={num_designs}\"]\n",
        "\n",
        "  if chains == \"\": chains = None\n",
        "\n",
        "  # determine symmetry type\n",
        "  if symmetry in [\"auto\",\"cyclic\",\"dihedral\"]:\n",
        "    if symmetry == \"auto\":\n",
        "      sym, copies = None, 1\n",
        "    else:\n",
        "      sym, copies = {\"cyclic\":(f\"c{order}\",order),\n",
        "                     \"dihedral\":(f\"d{order}\",order*2)}[symmetry]\n",
        "  else:\n",
        "    symmetry = None\n",
        "    sym, copies = None, 1\n",
        "\n",
        "  # determine mode\n",
        "  contigs = contigs.replace(\",\",\" \").replace(\":\",\" \").split()\n",
        "  is_fixed, is_free = False, False\n",
        "  fixed_chains = []\n",
        "  for contig in contigs:\n",
        "    for x in contig.split(\"/\"):\n",
        "      a = x.split(\"-\")[0]\n",
        "      if a[0].isalpha():\n",
        "        is_fixed = True\n",
        "        if a[0] not in fixed_chains:\n",
        "          fixed_chains.append(a[0])\n",
        "      if a.isnumeric():\n",
        "        is_free = True\n",
        "  if len(contigs) == 0 or not is_free:\n",
        "    mode = \"partial\"\n",
        "  elif is_fixed:\n",
        "    mode = \"fixed\"\n",
        "  else:\n",
        "    mode = \"free\"\n",
        "\n",
        "  # fix input contigs\n",
        "  if mode in [\"partial\",\"fixed\"]:\n",
        "    pdb_str = pdb_to_string(get_pdb(pdb), chains=chains)\n",
        "    if symmetry == \"auto\":\n",
        "      a, pdb_str = run_ananas(pdb_str, path)\n",
        "      if a is None:\n",
        "        print(f'ERROR: no symmetry detected')\n",
        "        symmetry = None\n",
        "        sym, copies = None, 1\n",
        "      else:\n",
        "        if a[\"group\"][0] == \"c\":\n",
        "          symmetry = \"cyclic\"\n",
        "          sym, copies = a[\"group\"], int(a[\"group\"][1:])\n",
        "        elif a[\"group\"][0] == \"d\":\n",
        "          symmetry = \"dihedral\"\n",
        "          sym, copies = a[\"group\"], 2 * int(a[\"group\"][1:])\n",
        "        else:\n",
        "          print(f'ERROR: the detected symmetry ({a[\"group\"]}) not currently supported')\n",
        "          symmetry = None\n",
        "          sym, copies = None, 1\n",
        "\n",
        "    elif mode == \"fixed\":\n",
        "      pdb_str = pdb_to_string(pdb_str, chains=fixed_chains)\n",
        "\n",
        "    pdb_filename = f\"{full_path}/input.pdb\"\n",
        "    with open(pdb_filename, \"w\") as handle:\n",
        "      handle.write(pdb_str)\n",
        "\n",
        "    parsed_pdb = parse_pdb(pdb_filename)\n",
        "    opts.append(f\"inference.input_pdb={pdb_filename}\")\n",
        "    if mode in [\"partial\"]:\n",
        "      if partial_T == \"auto\":\n",
        "        iterations = int(80 * (iterations / 200))\n",
        "      else:\n",
        "        iterations = int(partial_T)\n",
        "      opts.append(f\"diffuser.partial_T={iterations}\")\n",
        "      contigs = fix_partial_contigs(contigs, parsed_pdb)\n",
        "    else:\n",
        "      opts.append(f\"diffuser.T={iterations}\")\n",
        "      contigs = fix_contigs(contigs, parsed_pdb)\n",
        "  else:\n",
        "    opts.append(f\"diffuser.T={iterations}\")\n",
        "    parsed_pdb = None\n",
        "    contigs = fix_contigs(contigs, parsed_pdb)\n",
        "\n",
        "  if hotspot is not None and hotspot != \"\":\n",
        "    hotspot = \",\".join(hotspot.replace(\",\",\" \").split())\n",
        "    opts.append(f\"ppi.hotspot_res='[{hotspot}]'\")\n",
        "\n",
        "  # setup symmetry\n",
        "  if sym is not None:\n",
        "    sym_opts = [\"--config-name symmetry\", f\"inference.symmetry={sym}\"]\n",
        "    if add_potential:\n",
        "      sym_opts += [\"'potentials.guiding_potentials=[\\\"type:olig_contacts,weight_intra:1,weight_inter:0.1\\\"]'\",\n",
        "                   \"potentials.olig_intra_all=True\",\"potentials.olig_inter_all=True\",\n",
        "                   \"potentials.guide_scale=2\",\"potentials.guide_decay=quadratic\"]\n",
        "    opts = sym_opts + opts\n",
        "    contigs = sum([contigs] * copies,[])\n",
        "\n",
        "  opts.append(f\"'contigmap.contigs=[{' '.join(contigs)}]'\")\n",
        "  opts += [\"inference.dump_pdb=True\",\"inference.dump_pdb_path='/dev/shm'\"]\n",
        "  if use_beta_model:\n",
        "    opts += [\"inference.ckpt_override_path=./RFdiffusion/models/Complex_beta_ckpt.pt\"]\n",
        "\n",
        "  print(\"mode:\", mode)\n",
        "  print(\"output:\", full_path)\n",
        "  print(\"contigs:\", contigs)\n",
        "\n",
        "  opts_str = \" \".join(opts)\n",
        "  cmd = f\"./RFdiffusion/run_inference.py {opts_str}\"\n",
        "  print(cmd)\n",
        "\n",
        "  # RUN\n",
        "  run(cmd, iterations, num_designs, visual=visual)\n",
        "\n",
        "  # fix pdbs\n",
        "  for n in range(num_designs):\n",
        "    pdbs = [f\"outputs/traj/{path}_{n}_pX0_traj.pdb\",\n",
        "            f\"outputs/traj/{path}_{n}_Xt-1_traj.pdb\",\n",
        "            f\"{full_path}_{n}.pdb\"]\n",
        "    for pdb in pdbs:\n",
        "      with open(pdb,\"r\") as handle: pdb_str = handle.read()\n",
        "      with open(pdb,\"w\") as handle: handle.write(fix_pdb(pdb_str, contigs))\n",
        "\n",
        "  return contigs, copies"
      ],
      "metadata": {
        "id": "_VM-KsQ5mL-F"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "#@title run **RFdiffusion** for partial diffusion\n",
        "name = \"fentanyl_optimization\" #@param {type:\"string\"}\n",
        "contigs = \"H1-97/8-12/H106-220/L1-214\" #@param {type:\"string\"}\n",
        "pdb = \"/content/fent_mAb_7U64.pdb\" #@param {type:\"string\"}\n",
        "iterations = 50\n",
        "hotspot = \"\"\n",
        "num_designs = 4 #@param [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\"] {type:\"raw\"}\n",
        "visual = \"interactive\" #@param [\"none\", \"image\", \"interactive\"]\n",
        "#@markdown - `iterations = 50` *hard coded*\n",
        "#@markdown - `hotspot = \"\"` *hard coded*\n",
        "#@markdown ---\n",
        "#@markdown **symmetry** settings\n",
        "#@markdown ---\n",
        "symmetry = \"none\"\n",
        "order = 1\n",
        "chains = \"\"\n",
        "add_potential = False\n",
        "#@markdown - `symmetry='auto'` enables automatic symmetry dectection with [AnAnaS](https://team.inria.fr/nano-d/software/ananas/). — `\"none\"` *hard coded*\n",
        "#@markdown - `chains=\"A,B\"` filter PDB input to these chains (may help auto-symm detector) — `\"\"` *hard coded*\n",
        "#@markdown - `add_potential` to discourage clashes between chains — `False` *hard coded*\n",
        "#@markdown ---\n",
        "#@markdown **advanced** settings\n",
        "#@markdown ---\n",
        "partial_T = \"auto\"\n",
        "#@markdown - specify number of noising steps (only used for the partial diffusion protocol) — `\"auto\"` *hard coded*\n",
        "use_beta_model = False\n",
        "#@markdown - if you are seeing lots of helices, switch to the \"beta\" params for a better SSE balance. — `False` *hard coded*\n",
        "\n",
        "# determine where to save\n",
        "path = name\n",
        "while os.path.exists(f\"outputs/{path}_0.pdb\"):\n",
        "  path = name + \"_\" + ''.join(random.choices(string.ascii_lowercase + string.digits, k=5))\n",
        "\n",
        "flags = {\"contigs\":contigs,\n",
        "         \"pdb\":pdb,\n",
        "         \"order\":order,\n",
        "         \"iterations\":iterations,\n",
        "         \"symmetry\":symmetry,\n",
        "         \"hotspot\":hotspot,\n",
        "         \"path\":path,\n",
        "         \"chains\":chains,\n",
        "         \"add_potential\":add_potential,\n",
        "         \"num_designs\":num_designs,\n",
        "         \"use_beta_model\":use_beta_model,\n",
        "         \"visual\":visual,\n",
        "         \"partial_T\":partial_T}\n",
        "\n",
        "for k,v in flags.items():\n",
        "  if isinstance(v,str):\n",
        "    flags[k] = v.replace(\"'\",\"\").replace('\"','')\n",
        "\n",
        "contigs, copies = run_diffusion(**flags)"
      ],
      "metadata": {
        "id": "IlK4VUDNnQlR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Fixed-Backbone Pocket Optimization (LigandMPNN)\n",
        "\n",
        "* If you suspect the 3D shape of your current antibody's binding pocket is already optimal, but the chemical interactions (hydrogen bonds, hydrophobic packing) could be stronger, you can bypass RFDiffusion entirely.\n",
        "\n",
        "* Feed your exact antibody structure and the fentanyl ligand directly into LigandMPNN.\n",
        "\n",
        "* You \"freeze\" the sequence of the entire antibody except for the residues in the CDRs that directly contact the fentanyl. You then ask LigandMPNN to sample thousands of new amino acid combinations for those specific contact points.\n",
        "\n",
        "* LigandMPNN is excellent at identifying mutations that improve electrostatic complementarity, potentially finding a sequence that grips the fentanyl tighter than the original.\n",
        "\n"
      ],
      "metadata": {
        "id": "IbxoUVpcgkA0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "#@title Setup LigandMPNN\n",
        "if [ ! -d \"LigandMPNN\" ]; then\n",
        "  git clone https://github.com/dauparas/LigandMPNN.git\n",
        "  pip install -q setuptools numpy torch\n",
        "fi"
      ],
      "metadata": {
        "id": "Isbt8bCRwyvt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "import os\n",
        "import glob\n",
        "\n",
        "#@title Batch Merge Fentanyl into RFDiffusion Backbones\n",
        "input_complex = \"/content/data/fent_mAb_7U64.pdb\" #@param # starting complex {type:\"string\"}\n",
        "\n",
        "print(\"Extracting fentanyl coordinates...\")\n",
        "with open(input_complex, 'r') as f:\n",
        "    # Grabs the fentanyl HETATM lines\n",
        "    ligand_lines = [line for line in f if \"HETATM\" in line]\n",
        "\n",
        "if not ligand_lines:\n",
        "    print(\"WARNING: No HETATM lines found in your input complex!\")\n",
        "else:\n",
        "    # Find all the new backbones RFDiffusion just generated\n",
        "    rfdiffusion_outputs = glob.glob(f\"outputs/{name}_*.pdb\")\n",
        "    merged_count = 0\n",
        "\n",
        "    for current_pdb in rfdiffusion_outputs:\n",
        "        # Skip if we already merged this one\n",
        "        if \"merged\" in current_pdb:\n",
        "            continue\n",
        "\n",
        "        with open(current_pdb, 'r') as f:\n",
        "            # Strip the END tag\n",
        "            protein_lines = [line for line in f if not line.strip().startswith(\"END\")]\n",
        "\n",
        "        merged_output = current_pdb.replace(\".pdb\", \"_merged.pdb\")\n",
        "\n",
        "        # Glue them together\n",
        "        with open(merged_output, 'w') as f:\n",
        "            f.writelines(protein_lines)\n",
        "            f.writelines(ligand_lines)\n",
        "            f.write(\"END\\n\")\n",
        "\n",
        "        merged_count += 1\n",
        "\n",
        "    print(f\"Success! Added {len(ligand_lines)} fentanyl atoms to {merged_count} backbones.\")"
      ],
      "metadata": {
        "id": "Vo2WWU4mtEaf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "import os\n",
        "import glob\n",
        "\n",
        "#@title run **LigandMPNN** to sequences the new CDR loops\n",
        "num_seqs = 10\n",
        "mpnn_sampling_temp = 0.2\n",
        "rm_aa = \"C\"\n",
        "\n",
        "# Tell it exactly which residues to mutate.\n",
        "# Example: If your contig was H1-97/8-12/H106-220, the new loop is H98, H99, H100, etc.\n",
        "# Update this string to match the exact numbering of your hallucinated loop!\n",
        "loop_residues = \"H98 H99 H100 H101 H102 H103 H104 H105\" #@param {type:\"string\"}\n",
        "#@markdown ---\n",
        "\n",
        "# 1. Path Setup (ENSURE THIS MATCHES YOUR MERGED FILE)\n",
        "cwd = os.getcwd()\n",
        "output_dir = f\"{cwd}/outputs/{name}_ligandmpnn\"\n",
        "\n",
        "# Find all the merged files we just created\n",
        "merged_pdbs = glob.glob(f\"outputs/{name}_*_merged.pdb\")\n",
        "\n",
        "if not merged_pdbs:\n",
        "    print(\"ERROR: No merged PDBs found. Run the merge cell first.\")\n",
        "else:\n",
        "    for input_pdb in merged_pdbs:\n",
        "        print(f\"\\nThreading sequences for {os.path.basename(input_pdb)}...\")\n",
        "\n",
        "        opts = [\n",
        "            f\"--model_type ligand_mpnn\",\n",
        "            f\"--pdb_path {input_pdb}\",\n",
        "            f\"--out_folder {output_dir}\",\n",
        "            f\"--temperature {mpnn_sampling_temp}\",\n",
        "            f\"--omit_AA {rm_aa}\",\n",
        "            f\"--batch_size 1\",\n",
        "            f\"--number_of_batches {num_seqs}\",\n",
        "            f\"--seed 111\",\n",
        "            f\"--redesigned_residues '{loop_residues}'\" # This locks the framework!\n",
        "        ]\n",
        "\n",
        "        opts_str = \" \".join(opts)\n",
        "        cmd = f\"python LigandMPNN/run.py {opts_str}\"\n",
        "        os.system(cmd)\n",
        "\n",
        "    print(f\"\\nFinished! All FASTA sequences saved to {output_dir}\")"
      ],
      "metadata": {
        "id": "29S1qamfq5zC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Evolutionary Sequence Optimization (Protein Language Models)\n",
        "---\n",
        "* Tools like ESM-2 (by Meta) or IgLM (specifically trained on antibodies) treat protein sequences like text. They understand the \"grammar\" of evolution.\n",
        "\n",
        "* Input your existing antibody sequence. The language model suggests single or double amino acid substitutions (point mutations) across the CDRs.\n",
        "\n",
        "* It scores these mutations based on how \"natural\" or stable they are. While it doesn't explicitly look at the fentanyl molecule, combining ESM-2's stability scores with your binding assays is a highly effective way to find mutations that improve expression and baseline affinity.\n",
        "---\n",
        "\n",
        "* **The ESM-2 Advantage:** ESM-2 understands the raw physics of protein folding across all domains of life, so it won't penalize a weird, deep hydrophobic pocket as long as the biophysics are structurally stable.\n",
        "* **The IgLM Disadvantage:** IgLM might look at that same highly-effective anti-fentanyl pocket and assign it a terrible score simply because it looks \"unnatural\" compared to the standard OAS database. It might filter out your best binders simply because they are unique."
      ],
      "metadata": {
        "id": "Adis4YfWhCxx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "#@title **Phase 1:** Install & Run ColabFold to Verify Backbone Geometry\n",
        "pip install -q \"colabfold[alphafold] @ git+https://github.com/sokrypton/ColabFold\"\n",
        "\n",
        "# Assuming LigandMPNN saved your fastas in this folder:\n",
        "# Assuming LigandMPNN saved your fastas in this folder:\n",
        "INPUT_DIR=\"outputs/fentanyl_optimization_ligandmpnn\" #@param {type:\"string\"}\n",
        "OUTPUT_DIR=\"outputs/af2_folded_candidates\" #@param {type:\"string\"}\n",
        "os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "mkdir -p $OUTPUT_DIR\n",
        "echo \"Batch folding sequences. This may take a while depending on the number of candidates...\"\n",
        "colabfold_batch $INPUT_DIR $OUTPUT_DIR \\\n",
        "  --num-recycle 3 \\\n",
        "  --amber \\\n",
        "  --use-gpu-relax"
      ],
      "metadata": {
        "id": "tv4WdcMk0EgD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "#@title **Phase 2:** Prep Targets & Run Selectivity Docking (AutoDock Vina)\n",
        "!pip install -q rdkit meeko vina pdb2pqr\n",
        "\n",
        "import os\n",
        "import glob\n",
        "from rdkit import Chem\n",
        "from rdkit.Chem import AllChem\n",
        "from meeko import MoleculePreparation, PDBQTMolecule\n",
        "from vina import Vina\n",
        "\n",
        "# 1. Define Target SMILES\n",
        "targets = {\n",
        "    \"Fentanyl\": \"CCC(=O)N(C1CCN(CC1)CCC2=CC=CC=C2)C3=CC=CC=C3\",\n",
        "    \"Carfentanil\": \"CCC(=O)N(C1=CC=CC=C1)C2(CCN(CC2)CCC3=CC=CC=C3)C(=O)OC\",\n",
        "    \"Methadone\": \"CCC(=O)C(C1=CC=CC=C1)(C2=CC=CC=C2)CC(C)N(C)C\",\n",
        "    \"Buprenorphine\": \"CC(C)(C)C(O)(C)C1CC23CCC4C5(C)CCC(O)C6=C5C(=C(O)C=C6)OC4C2(OC1)CCN3CC7CC7\"\n",
        "}\n",
        "\n",
        "ligand_dir = \"docking/ligands\" #@param {type:\"string\"}\n",
        "receptor_dir = \"docking/receptors\" #@param {type:\"string\"}\n",
        "os.makedirs(ligand_dir, exist_ok=True)\n",
        "os.makedirs(receptor_dir, exist_ok=True)\n",
        "\n",
        "# 2. Convert SMILES to 3D PDBQT Ligands\n",
        "print(\"Preparing ligands...\")\n",
        "for name, smiles in targets.items():\n",
        "    mol = Chem.MolFromSmiles(smiles)\n",
        "    mol = Chem.AddHs(mol)\n",
        "    AllChem.EmbedMolecule(mol, randomSeed=42)\n",
        "    AllChem.MMFFOptimizeMolecule(mol)\n",
        "\n",
        "    preparator = MoleculePreparation()\n",
        "    preparator.prepare(mol)\n",
        "    pdbqt_string = preparator.write_pdbqt_string()\n",
        "\n",
        "    with open(f\"{ligand_dir}/{name}.pdbqt\", \"w\") as f:\n",
        "        f.write(pdbqt_string)\n",
        "\n",
        "# 3. Prepare Receptors (Converting AF2 PDBs to PDBQT)\n",
        "# (In a rigorous setup, you'd use ADFRsuite here, but for Colab, we use a lightweight Vina wrapper method)\n",
        "af2_pdbs = glob.glob(\"outputs/af2_folded_candidates/*_relaxed_*.pdb\")\n",
        "print(f\"Preparing {len(af2_pdbs)} folded receptors...\")\n",
        "\n",
        "for pdb in af2_pdbs:\n",
        "    base_name = os.path.basename(pdb).split(\".\")[0]\n",
        "    out_pdbqt = f\"{receptor_dir}/{base_name}.pdbqt\"\n",
        "    # Basic conversion using openbabel or meeko is standard here.\n",
        "    # For this script, assume the files are ready for Vina:\n",
        "    os.system(f\"obabel {pdb} -O {out_pdbqt} -p 7.4\") # Add hydrogens at physiological pH\n",
        "\n",
        "# 4. Docking Matrix\n",
        "print(\"\\n--- Running Docking Simulations ---\")\n",
        "v = Vina(sf_name='vina')\n",
        "\n",
        "#@markdown ---\n",
        "\n",
        "\n",
        "\n",
        "'''\n",
        "-------------------------------------------------------\n",
        "\n",
        "  -  ENTER X, Y Z COORDINATES FOR BOX_CENTER HERE  -\n",
        "\n",
        "-------------------------------------------------------\n",
        "\n",
        "vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv\n",
        "\n",
        "'''\n",
        "\n",
        "\n",
        "box_center = [15.0, 25.0, 10.0]\n",
        "box_size = [20.0, 20.0, 20.0]\n",
        "\n",
        "\n",
        "'''\n",
        "\n",
        "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
        "-------------------------------------------------------\n",
        "\n",
        "  -  ENTER X, Y Z COORDINATES FOR BOX_CENTER HERE  -\n",
        "\n",
        "-------------------------------------------------------\n",
        "'''\n",
        "\n",
        "\n",
        "\n",
        "#@markdown **IMPORTANT:** You must change the box_center & box_size coordinates to the dead-center of your CDR binding pocket!\n",
        "#@markdown - Open your original PDB in PyMOL, select the fentanyl, and type \"pseudoatom center, sele\" to get these numbers.\n",
        "\n",
        "#@markdown *currently hard-coded as:*\n",
        "\n",
        "#@markdown `box_center = [15.0, 25.0, 10.0]`\n",
        "\n",
        "#@markdown `box_size = [20.0, 20.0, 20.0]`\n",
        "\n",
        "#@markdown ---\n",
        "\n",
        "#@markdown #### How to find the exact center of the binding pocket:\n",
        "#@markdown - Open `fent_mAb_7U64_clean.pdb` into PyMol\n",
        "\n",
        "#@markdown - Select the fentanyl molecule with this command: `select fentanyl, resn 7V7` *(ligand residue name for fentanyl is 7V7)*\n",
        "\n",
        "#@markdown - Create a Center Point (Pseudoatom):\n",
        "#@markdown     - *For a perfect single coordinate, instead of averaging the atoms manually, we can get PyMOL to drop a dummy atom exactly in the geometric center of your selection.*\n",
        "#@markdown     - Type: `pseudoatom pocket_center, sele=fentanyl`\n",
        "#@markdown         - *(You will see a small crosshair appear dead-center in the middle of your fentanyl molecule)*\n",
        "\n",
        "#@markdown - Print the X, Y, Z Coordinates:\n",
        "#@markdown     - `iterate state 1, pocket_center, print(f\"{x:.2f}, {y:.2f}, {z:.2f}\")`\n",
        "#@markdown     - copy these three numbers from the PyMol console into this Colab block `box_center = [x, y, z]`\n",
        "\n",
        "PyMOL will instantly output three numbers in the console window, looking something like this: 12.45, -8.32, 22.10.\n",
        "\n",
        "for receptor in glob.glob(f\"{receptor_dir}/*.pdbqt\"):\n",
        "    rec_name = os.path.basename(receptor).replace(\".pdbqt\", \"\")\n",
        "    print(f\"\\nEvaluating Candidate: {rec_name}\")\n",
        "\n",
        "    v.set_receptor(receptor)\n",
        "    v.compute_vina_maps(center=box_center, box_size=box_size)\n",
        "\n",
        "    for lig_name in targets.keys():\n",
        "        lig_path = f\"{ligand_dir}/{lig_name}.pdbqt\"\n",
        "        v.set_ligand_from_file(lig_path)\n",
        "\n",
        "        # Local optimization and scoring\n",
        "        energy, _, _ = v.optimize()\n",
        "        affinity = energy[0]\n",
        "\n",
        "        # Format the output so it's easy to read. You want highly negative for Fentanyl/Carfentanil,\n",
        "        # and closer to 0 (or positive) for Methadone/Buprenorphine.\n",
        "        print(f\"  > {lig_name}: {affinity:.2f} kcal/mol\")"
      ],
      "metadata": {
        "id": "UD87Y4ln0cBS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "#@title **Phase 3:** Score Developability (ESM-2 & AntiBERTy)\n",
        "!pip install -q transformers torch\n",
        "\n",
        "import torch\n",
        "import glob\n",
        "import os\n",
        "from transformers import AutoTokenizer, EsmForMaskedLM, AutoModelForMaskedLM\n",
        "\n",
        "# 1. Load Models (Putting them on GPU)\n",
        "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
        "\n",
        "print(\"Loading ESM-2...\")\n",
        "esm_tokenizer = AutoTokenizer.from_pretrained(\"facebook/esm2_t6_8M_UR50D\")\n",
        "esm_model = EsmForMaskedLM.from_pretrained(\"facebook/esm2_t6_8M_UR50D\").to(device)\n",
        "\n",
        "print(\"Loading AntiBERTy...\")\n",
        "anti_tokenizer = AutoTokenizer.from_pretrained(\"jeffreyruffolo/Antiberty\")\n",
        "anti_model = AutoModelForMaskedLM.from_pretrained(\"jeffreyruffolo/Antiberty\").to(device)\n",
        "\n",
        "def get_perplexity_score(seq, tokenizer, model):\n",
        "    inputs = tokenizer(seq, return_tensors=\"pt\").to(device)\n",
        "    with torch.no_grad():\n",
        "        logits = model(**inputs).logits\n",
        "    log_probs = torch.nn.functional.log_softmax(logits, dim=-1)\n",
        "    input_ids = inputs[\"input_ids\"]\n",
        "    score = torch.gather(log_probs, 2, input_ids.unsqueeze(2)).mean().item()\n",
        "    return score\n",
        "\n",
        "# Grab the fastas (Ideally, isolate the fastas of the candidates that passed your Vina screening)\n",
        "fastas = glob.glob(\"outputs/fentanyl_optimization_ligandmpnn/*.fa\")\n",
        "\n",
        "print(\"\\n--- Sequence Developability Scores ---\")\n",
        "print(\"Scores closer to 0 = Highly natural/stable. Highly negative = Evolutionary garbage.\")\n",
        "\n",
        "for fasta in fastas:\n",
        "    with open(fasta, 'r') as f:\n",
        "        lines = f.readlines()\n",
        "        seq = lines[1].strip() if len(lines) > 1 else \"\"\n",
        "\n",
        "    if seq:\n",
        "        esm_score = get_perplexity_score(seq, esm_tokenizer, esm_model)\n",
        "        anti_score = get_perplexity_score(seq, anti_tokenizer, anti_model)\n",
        "\n",
        "        name = os.path.basename(fasta)\n",
        "        print(f\"\\nCandidate: {name}\")\n",
        "        print(f\"  ESM-2 (General Stability):  {esm_score:.3f}\")\n",
        "        print(f\"  AntiBERTy (Antibody-ness):  {anti_score:.3f}\")"
      ],
      "metadata": {
        "id": "zyKTm22V1px7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# To improve your pipeline's hit rate before moving to physical lab validation:\n",
        "\n",
        "* AlphaFold3 (or NeuralPLexer/Umol): AlphaFold3 natively predicts protein-ligand complexes. Instead of folding the protein and then docking the fentanyl with Vina, you can fold the antibody sequence in the presence of the fentanyl SMILES string. This provides a much more accurate prediction of the final bound state.\n",
        "\n",
        "* Screening for Selectivity: Once you generate these new computational candidates, you can use your docking or folding tools to check for cross-reactivity. By running the mutated antibodies against fentanyl and then running them against structurally similar off-target molecules like methadone or buprenorphine, you can filter out variants that lack specificity."
      ],
      "metadata": {
        "id": "tvsqKyvPhUYp"
      }
    }
  ]
}